### PR TITLE
[BUG] Drop annotations with NaN onset in EEGLAB raw files 

### LINKED ISSUE #12425: Annotation-related error reading EEGLAB file
### Description of the problem

When trying to read a specific EEGLAB file, there is an error related to annotations and absolute onset time. This problem was reported in our forum: https://mne.discourse.group/t/read-raw-eeglab-gives-valueerror-cannot-convert-float-nan-to-integer/8285

### Steps to reproduce

```Python
import openneuro as on

import mne

on.download(
    dataset="ds004842",
    include=[
        "sub-001/**/*run-1*.set",
        "sub-001/**/*run-1*.fdt",
    ],
)

raw = mne.io.read_raw_eeglab(
    "ds004842/sub-001/ses-C1C/eeg/sub-001_ses-C1C_task-DriveOnMission_run-1_eeg.set"
)
```

### Link to data

https://openneuro.org/datasets/ds004842/versions/1.0.0

### Expected results

I should be able to import the data.

### Actual results

There is an error when importing the data:

```
<stdin>:1: RuntimeWarning: Unknown types found, setting as type EEG:
inferiorveog: ['EXG4']
leftheog: ['EXG1']
leftmastoid: ['EXG5']
rightheog: ['EXG2']
rightmastoid: ['EXG6']
superiorveog: ['EXG3']
Reading /Users/clemens/Projects/mne-python/ds004842/sub-001/ses-C1C/eeg/sub-001_ses-C1C_task-DriveOnMission_run-1_eeg.fdt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/clemens/Projects/mne-python/mne/io/eeglab/eeglab.py", line 332, in read_raw_eeglab
    return RawEEGLAB(
           ^^^^^^^^^^
  File "<decorator-gen-238>", line 12, in __init__
  File "/Users/clemens/Projects/mne-python/mne/io/eeglab/eeglab.py", line 505, in __init__
    self.set_annotations(annot)
  File "<decorator-gen-225>", line 12, in set_annotations
  File "/Users/clemens/Projects/mne-python/mne/io/base.py", line 734, in set_annotations
    new_annotations.crop(
  File "<decorator-gen-38>", line 12, in crop
  File "/Users/clemens/Projects/mne-python/mne/annotations.py", line 636, in crop
    absolute_onset = timedelta(seconds=onset) + offset
                     ^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: cannot convert float NaN to integer
```

### Additional information

```
Platform             macOS-14.3-arm64-arm-64bit
Python               3.12.1 (main, Jan 24 2024, 11:32:51) [Clang 15.0.0 (clang-1500.1.0.2.5)]
Executable           /Users/clemens/.pyenv/versions/3.12.1/envs/mne/bin/python3.12
CPU                  arm (12 cores)
Memory               32.0 GB

Core
├☑ mne               1.7.0.dev77+g566c6ea5a (devel, latest release is 1.6.1)
├☑ numpy             1.26.3 (OpenBLAS 0.3.23.dev with 12 threads)
├☑ scipy             1.12.0
└☑ matplotlib        3.8.2 (backend=MacOSX)

Numerical (optional)
└☐ unavailable       sklearn, numba, nibabel, nilearn, dipy, openmeeg, cupy, pandas

Visualization (optional)
├☑ qtpy              2.4.1 (PySide6=6.6.1)
├☑ pyqtgraph         0.13.3
├☑ mne-qt-browser    0.6.1
└☐ unavailable       pyvista, pyvistaqt, vtk, ipympl, ipywidgets, trame_client, trame_server, trame_vtk, trame_vuetify

Ecosystem (optional)
└☐ unavailable       mne-bids, mne-nirs, mne-features, mne-connectivity, mne-icalabel, mne-bids-pipeline, neo
```

### PR DESCRIPTION
**Pull request by withmywoessner**
#### Reference issue
Fixes #12425.

#### What does this implement/fix?
Fix bug where MNE cannot handle EEGLAB events with NaN annotations

#### Additional information
Rationale for decision from EEGLAB documentation:
image
https://eeglab.org/tutorials/04_Import/Importing_Event_Epoch_Info.html

### CODE REVIEW

**Comment by withmywoessner** [comment]
@cbrnr Should be fixed now!

**Comment by cbrnr** [comment]
I would not add a new dataset, but can you add a test with some artificially generated markers that demonstrate the problem (and verify the fix)?

**Comment by withmywoessner** [comment]
> I would not add a new dataset, but can you add a test with some artificially generated markers that demonstrate the problem (and verify the fix)?

Would I have to make a temp file @cbrnr? Because from my understanding, I would have to read a MATLAB .set file then edit the annotations. Then I would need to read the file again to check to see if it drops the bad annotations. Can you think of another way?

**Comment by cbrnr** [comment]
Hm, yes, because the fix is in the reader it's probably easiest if you add a small(!) test file then.

**Comment by larsoner** [comment]
> EDIT3: Yeah, since MATLAB files are stored as a binary and the _read_annotations expects a MATLAB file. I think I would have to read the MATLAB file, then edit the annotations, then re-export it as a binary.

Can you try this and see if it's easy? If it's old format then `scipy.io` functions can be used. If it's new format then `h5py` could perhaps be used directly just to change one or two entries and save to a `tmpdir`?

**Comment by withmywoessner** [comment]
I can try @larsoner should I just use the `tempfile`library?

**Comment by larsoner** [comment]
No in tests typically you use the pytest test fixture `def test_something(tmp_path):` and write for example in the test `with open(tmp_path / "my_file.txt', 'w'):` or similar.

**Comment by withmywoessner** [comment]
Should be good now @larsoner and @cbrnr. Thanks!

**Comment by withmywoessner** [comment]
There is an error related to testing. Not sure what the problem is. When I ran it locally it passed:
image

**Comment by larsoner** [comment]
That's a weird error -- I put a little fix in #12491 so ignore it for now and we'll merge that quickly and merge `main` into your branch afterward to make sure everything is green

File: mne/io/eeglab/tests/test_eeglab.py [hunk 1]
```diff
@@ -719,3 +720,39 @@ def get_bad_information(eeg, get_pos, *, montage_units):
     assert len(pos["lpa"]) == 3
     assert len(pos["rpa"]) == 3
     assert len(raw.info["dig"]) == n_eeg + 3
+
+
+@testing.requires_testing_data
+def test_eeglab_drop_nan_annotations(tmp_path):
+    """Test reading file with NaN annotations."""
```
Comment (larsoner) [inline mne/io/eeglab/tests/test_eeglab.py:727]:
```suggestion
    """Test reading file with NaN annotations."""
    pytest.importorskip("eeglabio")
```

File: mne/io/eeglab/tests/test_eeglab.py [hunk 2]
```diff
@@ -719,3 +720,40 @@ def get_bad_information(eeg, get_pos, *, montage_units):
...
+        ch_types=np.repeat("EEG", len(ch_names)),
+    )
+
+    with pytest.raises(
+        RuntimeWarning,
+        match=(
+            r"1 events have an onset that is NaN. These values are usually "
+            r"ignored by EEGLAB and will be dropped from the annotations."
+        ),
+    ):
```
Comment (larsoner) [inline mne/io/eeglab/tests/test_eeglab.py:758]: Usually it's best to make these as short as possible while still being selective, that way if we tweak wording later in the warning for example we don't need to update tests
```suggestion
    with pytest.raises(RuntimeWarning, match=r"1 events have an onset that is NaN.*"):
```

**Comment by larsoner** [comment]
Just a couple of minor things, committed and marking for merge-when-green, thanks @withmywoessner !

File: mne/io/eeglab/tests/test_eeglab.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 import numpy as np
 import pytest
+from eeglabio.raw import export_set
```
Comment (larsoner) [inline mne/io/eeglab/tests/test_eeglab.py:14]: Must be nested
```suggestion
```

File: mne/io/eeglab/tests/test_eeglab.py [hunk 4]
```diff
@@ -719,3 +720,34 @@ def get_bad_information(eeg, get_pos, *, montage_units):
     assert len(pos["lpa"]) == 3
     assert len(pos["rpa"]) == 3
     assert len(raw.info["dig"]) == n_eeg + 3
+
+
+@testing.requires_testing_data
+def test_eeglab_drop_nan_annotations(tmp_path):
+    """Test reading file with NaN annotations."""
+    pytest.importorskip("eeglabio")
```
Comment (larsoner) [inline mne/io/eeglab/tests/test_eeglab.py:728]:
```suggestion
    pytest.importorskip("eeglabio")
    from eeglabio.raw import export_set

```

**Comment by larsoner** [comment]
Pip-pre is unrelated, merging!

**Pull request merged**